### ISSUE TITLE
Pupil mask value is not present in `psf_grid` output

### ISSUE THREAD

**Issue by bhilbert4** [spacetelescope/webbpsf#292]
I've been thinking about the easiest way to allow Mirage users to specify fits files containing PSF grids. The two options I've been thinking about involve a rigid filename format that includes all relevant instrument and webbpsf details (detector/filter/pupil/oversampling/OPD/etc), or having Mirage look through the headers of all fits files in a given directory to identify the proper file.

But in the course of producing gridded PSF files, I've found that the value of the pupil mask used by WebbPSF is not present in the output file header, nor in the (automatically generated) filename. For example, in the code below, running with and without the `nrc.pupil_mask` line produces files with the same name and header info, but different PSFs.

```
nrc = webbpsf.NIRCam() 
nrc.filter = 'F200W' 
nrc.detector = 'NRCB1' 
nrc.pupil_mask = 'WEAK LENS +8'                                                                                                                                                                   

grid = nrc.psf_grid(num_psfs=4, all_detectors=False, oversample=2, fov_pixels=301, 
                                add_distortion=False, save=True) 
```

**Comment by bhilbert4** [comment]
I should also point out that the OPD realization number is also not present in the file header nor filename.

**Comment by shanosborne** [comment]
For both of these comments, it was a conscious choice not to include the information in the default file name, because at some point it just becomes too long (and if people want it they can use the `outfile` keyword arg to change the file name). But I agree that this information should be in the header/meta dictionary. Now as to whether or not they’re in there:
* For your first comment, yes you’re right this is a bug. I’ll add it so you’ll get `PUPIL` and `CORONMSK` keywords if you set the `pupil_mask` and `image_mask` respectively.
* For your second comment, this information is in fact in the header/meta dictionary. If you set your `pupilopd` to be a certain realization (e.g. `nrc.pupilopd = ('OPD_RevW_ote_for_NIRCam_predicted.fits.gz', 7)`), your `PUPILOPD` keyword will say `PUPILOPD= 'OPD_RevW_ote_for_NIRCam_predicted.fits.gz slice 7' / Pupil OPD source`. If you don’t specify the slice (so webbpsf defaults to the first slice), your  `PUPILOPD` keyword will just say `PUPILOPD= 'OPD_RevW_ote_for_NIRCam_requirements.fits.gz' / Pupil OPD source`. This part does require the knowledge that not specifying the slice will result in using the default first slice, but it also exactly matches what the header will look like if you were to create a PSF using `calc_psf`.

**Comment by bhilbert4** [comment]
One other comment with regard to pupil values:
I created some NIRISS libraries this afternoon, looping over filters and pupil_mask values. It didn't occur to me at the time, but since half of the NIRISS filters are in the pupil wheel, there are certain filter/pupil combinations that aren't possible, unless I'm mistaken. For example, combining the NRM with F200W (or any filters blueward of that). But `psf_grid` did produce a gridded PSF object for these cases without complaining. The output PSFs appear to be identical.

**Comment by mperrin** [comment]
Correct. WebbPSF does not (by design) try to model all instrument details such as when filters are actually in pupil wheels. Users are expected to be familiar with the capabilities of instruments, enough to know if they’re asking for simulations of an unsupported and perhaps not physically possible mode. 

This is on purpose both because it would be more work to raise alarms if users ask for unsupported modes, and also because sometimes there are unsupported modes that are still of interest for calibration purposes (for instance NIRspec Imaging) or because they might later become supported modes (looking at NIRCam LW coronagraph spots with SW filters, or vice versatility). But you’re right there are some cases that just aren’t achievable given instrument hardware.

**Comment by mperrin** [comment]
Note there's actually a subtlety here with NIRISS. It has an "auto pupil" feature where if you assign the filter, it will automatically toggle to the right pupil (CLEAR or CLEARP) depending on which wheel the pupil is. The intent is to make it easy & automatic for users to get the right pupil in regular imaging mode. If you set `pupil_mask='MASK_NRM'` first and then iterate through filters, I bet you engage that feature and it will automatically toggle to CLEAR. On the other hand if you set the filter and then explicitly set the pupil mask to NRM, then you can force webbpsf into simulating the non-achievable combination, like I was saying.

**Comment by shanosborne** [comment]
The two issues brought up in this issue have now both been fixed - The pupil_mask not being included was fixed in PR #293 and the OPD realization not being included was fixed in PR #299, so I'm going to close this issue unless there are any other concerns.
